File: 93_balanced-binary-tree/balanced-binary-tree.java
/*
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/balanced-binary-tree
@Language: Java
@Datetime: 16-05-18 13:44
*/

/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: True if this Binary tree is Balanced, or false.
     */
    public boolean isBalanced(TreeNode root) {
        // write your code here
        if (root == null) {
            return true;
        }
        if (root.left == null && root.right != null) {
            if (root.right.left != null || root.right.right != null) {
                return false;
            } else {
                return true;
            }
        }
        if (root.right == null && root.left != null) {
            if (root.left.left != null || root.left.right != null) {
                return false;
            } else {
                return true;
            }
        }
        boolean left = isBalanced(root.left);
        boolean right = isBalanced(root.right);
        return left && right;
    }
}

File: 93_balanced-binary-tree/balanced-binary-tree.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/balanced-binary-tree
@Language: Python
@Datetime: 16-05-18 13:38
'''

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    """
    @param root: The root of binary tree.
    @return: True if this Binary tree is Balanced, or false.
    """
    def isBalanced(self, root):
        # write your code here
        if root is None:
            return True
        if root.left is None and root.right is not None:
            if root.right.left is not None or root.right.right is not None:
                return False
            return True
        if root.right is None and root.left is not None:
            if root.left.left is not None or root.left.right is not None:
                return False
            return True
        left = self.isBalanced(root.left)
        right = self.isBalanced(root.right)
        return left and right
        